src/key_extractor.rs

```diff
@@ -1,47 +1,219 @@
+use actix_http::StatusCode;
 use actix_web::{dev::ServiceRequest, http::header::ContentType};
+use actix_web::{HttpResponse, HttpResponseBuilder, ResponseError};
 use governor::clock::{Clock, DefaultClock, QuantaInstant};
 use governor::NotUntil;
 
-use std::{fmt::Display, hash::Hash, net::IpAddr};
+use std::fmt::{Debug, Display};
+use std::{hash::Hash, net::IpAddr};
 
 /// Generic structure of what is needed to extract a rate-limiting key from an incoming request.
+///
+/// ## Example
+/// ```rust
+/// use actix_governor::{KeyExtractor, SimpleKeyExtractionError};
+/// use actix_web::ResponseError;
+/// use actix_web::dev::ServiceRequest;
+///
+/// #[derive(Clone)]
+/// struct Foo;
+///
+/// // will return 500 error and 'Extract error' as content
+/// impl KeyExtractor for Foo {
+///     type Key = ();
+///     type KeyExtractionError = SimpleKeyExtractionError<&'static str>;
+///
+///     fn extract(&self, _req: &ServiceRequest) -> Result<Self::Key, Self::KeyExtractionError> {
+///         Err(SimpleKeyExtractionError::new("Extract error"))
+///     }
+/// }
+/// ```
+///
+/// For more see [`custom_key_bearer`](https://github.com/AaronErhardt/actix-governor/blob/main/examples/custom_key_bearer.rs) example
 pub trait KeyExtractor: Clone {
     /// The type of the key.
     type Key: Clone + Hash + Eq;
 
     /// The type of the error that can occur if key extraction from the request fails.
-    type KeyExtractionError: Display;
+    type KeyExtractionError: ResponseError + 'static;
 
     #[cfg(feature = "log")]
     /// Name of this extractor (only used in logs).
     fn name(&self) -> &'static str;
 
-    /// Extraction method
+    /// Extraction method, will return [`KeyExtractionError`] response when the extract failed
+    ///
+    /// [`KeyExtractionError`]: KeyExtractor::KeyExtractionError
     fn extract(&self, req: &ServiceRequest) -> Result<Self::Key, Self::KeyExtractionError>;
 
     /// The content you want to show it when the rate limit is exceeded.
     /// The [`NotUntil`] will be passed to it and it has enough information.
     /// You need to return the content and the content type.
```

Comment:
I think this still needs to be updated. Maybe something like this:

```md
You can calculate the time at which a caller can expect the next positive rate-limiting result by using [`NotUntil`].
The [`HttpResponseBuilder`] allows you to build a fully customized [`HttpResponse`] in case of an error.
```